waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/tunneling/TunnelingMetaStoreClientFactoryTest.java

```diff
@@ -0,0 +1,154 @@
+        .thenReturn(hiveMetaStoreClientSupplier);
+    when(localPortFactory.getLocalPort()).thenReturn(localPort);
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_PRIVATE_KEYS.varname, "private_key");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_KNOWN_HOSTS.varname, "");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_STRICT_HOST_KEY_CHECKING.varname, "yes");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_ROUTE.varname, "user@hop1 -> hop2");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_LOCALHOST.varname, localHost);
+
+    tunnelingMetaStoreClientFactory = new TunnelingMetaStoreClientFactory(tunnelableFactorySupplier,
+        metaStoreClientFactory, localHiveConfFactory, hiveMetaStoreClientSupplierFactory, localPortFactory);
+
+  }
+
+  @Test
+  public void newInstanceWithTunneling() throws Exception {
+    ArgumentCaptor<TunnelableSupplier> captor = ArgumentCaptor.forClass(TunnelableSupplier.class);
+    tunnelingMetaStoreClientFactory.newInstance(hiveConf, name, reconnectionRetries);
+    verify(tunnelableFactory)
+        .wrap(captor.capture(), eq(tunnelingMetaStoreClientFactory.METHOD_CHECKER), eq("my-machine"), anyInt(),
+            eq("localhost"), eq(metastore.getThriftPort()));
+  }
+
+  @Test
+  public void localHiveConfigUsesCorrectParameters() {
+    tunnelingMetaStoreClientFactory.newInstance(hiveConf, name, reconnectionRetries);
+    ArgumentCaptor<String> stringArgument = ArgumentCaptor.forClass(String.class);
+    ArgumentCaptor<Integer> intArgument = ArgumentCaptor.forClass(Integer.class);
+    verify(localHiveConfFactory).createLocalHiveConf(stringArgument.capture(), intArgument.capture(), eq(hiveConf));
+    assertThat(stringArgument.getValue(), is(localHost));
+    assertThat(intArgument.getValue(), is(localPort));
+  }
+
+  @Test
+  public void tunnelableFactorySupplierUsesCorrectParameters() {
+    tunnelingMetaStoreClientFactory.newInstance(hiveConf, name, reconnectionRetries);
+    verify(tunnelableFactorySupplier).get(eq(localHiveConf));
+  }
+
+  @Test
+  public void hiveMetaStoreClientSupplierFactoryUsesCorrectParameters() {
+    tunnelingMetaStoreClientFactory.newInstance(hiveConf, name, reconnectionRetries);
+    verify(hiveMetaStoreClientSupplierFactory).newInstance(eq(localHiveConf), eq(name), eq(reconnectionRetries));
+  }
+
+  @Test
+  public void tunnelableFactoryUsesCorrectParameters() throws URISyntaxException {
+    URI uri = new URI(metastore.getThriftConnectionUri());
+    String remoteHost = uri.getHost();
+    int remotePort = uri.getPort();
+    tunnelingMetaStoreClientFactory.newInstance(hiveConf, name, reconnectionRetries);
+    verify(tunnelableFactory)
+        .wrap(eq(hiveMetaStoreClientSupplier), eq(tunnelingMetaStoreClientFactory.METHOD_CHECKER), eq(localHost),
+            any(Integer.class), eq(remoteHost), eq(remotePort));
+  }
+
+  @Test
+  public void newInstanceWithoutTunneling() throws Exception {
```

Comment:
use *asserts* for validating your test if possible. `verify` is to test delegation you should only need to use it if you have some kind of method that doesn't return anything and you want to make sure something else is called correctly. 
In this case the class under test returns something, so all test should end with an assertion of the object.

example instead of:
```
class_under_test.create(someFactory);
verify(factory).methodCalled("string");
```
You mock the behavior of `factory` like so:
```
when(factory.methodCall("string").thenReturn("anotherString");
String result = class_under_test.create(someFactory);
assertThat(result, is("anotherString");
```

You are not validating the returned object at all which is bad as that's the thing that counts the most. Mock the expected behavior from the classes you inject, that documents what you expect from them better then `verify` they are called, its more important mock what they return.